<?xml version="1.0"?>
<launch>
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find mav_tunnel_nav)/config/rbpf/rosconsole.config"/>

  <arg name="mav_name" default="iris" />
  <arg name="x" default="0.0"/>
  <arg name="y" default="0.0"/>
  <arg name="z" default="0.21"/>

  <!-- Start Gazebo -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name"
         value="$(find mav_tunnel_nav)/worlds/IndianTunnel.world" />
    <!--
     <arg name="world_name"
          value="$(find rotors_gazebo)/worlds/basic.world" />
    -->
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
    <arg name="extra_gazebo_args" value="--verbose"/>
  </include>

  <group ns="iris">
    <include file="$(find rotors_gazebo)/launch/spawn_mav.launch">
      <arg name="mav_name" value="$(arg mav_name)" />
      <!--<arg name="model" value="$(find mav_tunnel_nav)/urdf/iris/iris_14ranges.xacro" />-->
      <arg name="model" value="$(find mav_tunnel_nav)/urdf/iris/iris_depthcam.xacro" />
      <arg name="enable_logging" value="false" />
      <arg name="enable_ground_truth" value="false" />
      <arg name="log_file" value="iris"/>
      <arg name="x" value="$(arg x)"/>
      <arg name="y" value="$(arg y)"/>
      <arg name="z" value="$(arg z)"/>
    </include>

    <!-- Converts joystick movements into sensors_msgs/Joy messages -->
    <node name="joy_node" pkg="joy" type="joy_node" />

    <!-- Converts sensor_msgs/Joy to mav_msgs/RollPitchYawRateThrust messages -->
    <node name="rotors_joy_interface" pkg="rotors_joy_interface" type="rotors_joy_interface">
      <!--
      <rosparam command="load" file="$(find mav_tunnel_nav)/config/joypads/logitech_f310.yaml" />
      -->
      <rosparam command="load" file="$(find mav_tunnel_nav)/config/joypads/sanwa.yaml" />
    </node>

    <!-- Sends motor commands based given commanded rates for roll/pitch/yaw and thrust -->
    <node name="roll_pitch_yawrate_thrust_controller_node" pkg="rotors_control" type="roll_pitch_yawrate_thrust_controller_node">
      <rosparam command="load" file="$(find rotors_gazebo)/resource/roll_pitch_yawrate_thrust_controller_iris.yaml" />
      <rosparam command="load" file="$(find rotors_gazebo)/resource/iris.yaml" />
      <remap from="odometry" to="imu_odom/odometry" />
    </node>

    <!-- imu odometry -->
    <node name="imu_odom" pkg="mav_tunnel_nav" type="imu_odom">
      <param name="imu_topic" value="imu"/>
      <param name="odom_topic" value="imu_odom/odometry"/>
      <param name="world_frame_id" value="world"/>
      <param name="odom_frame_id" value="odom"/>
    </node>

    <!-- rbpf -->
    <node name="rbpf" pkg="mav_tunnel_nav" type="rbpf" output="screen">
      <param name="odom_topic" value="imu_odom/odometry"/>
      <param name="pc_topic" value="camera_ir/camera/depth/points"/>
      <param name="world_frame_id" value="world"/>
      <param name="robot_frame_id" value="base_link"/>

      <param name="n_particles" value="30"/>
      <param name="update_freq" value="3"/>

      <param name="depth_cam_pc_downsample" value="1000"/>

      <param name="map_resol" value="0.1"/>
      <param name="map_probHit" value="0.8"/>
      <param name="map_probMiss" value="0.2"/>
      <param name="map_threshMin" value="0.12"/>
      <param name="map_threshMax" value="0.97"/>

      <param name="t_only_mapping" value="10.0"/>
      <param name="mapping_interval" value="3"/>
    </node>

    <!-- Given robot_description parameter and joint states, publish TF frames -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" />

    <node pkg="rviz" type="rviz" name="rviz_nav" output="screen" args="-d $(find mav_tunnel_nav)/config/rbpf/rbpf.rviz"></node>
  </group>
</launch>
